File: core-java/src/test/java/chessgame/domain/model/PawnWhiteTest.java
package chessgame.domain.model;

import chessgame.domain.Piece;
import org.junit.Before;
import org.junit.Test;

import java.util.Set;

import static chessgame.domain.factory.PieceFactory.createWhitePawn;
import static chessgame.domain.model.BoardPosition.*;
import static chessgame.domain.model.PieceColor.WHITE;
import static org.junit.Assert.*;

public class PawnWhiteTest {

    private static final BoardPosition INITIAL_POSITION = D2;

    private Board board;

    private Piece whitePawn;

    @Before
    public void setUp() {
        board = new Board();
        board.init();
        whitePawn = createWhitePawn(board, INITIAL_POSITION);
    }

    @Test
    public void createInstanceTest() {
        assertTrue(whitePawn instanceof Pawn);
    }

    @Test
    public void whiteColorTest() {
        assertEquals(WHITE, whitePawn.getColor());
    }

    @Test
    public void initPositionTest() {
        assertEquals(INITIAL_POSITION, whitePawn.getPosition());
        assertEquals(whitePawn, board.get(INITIAL_POSITION));
    }

    @Test
    public void moveToD4Test() {
        assertNotNull(board.get(INITIAL_POSITION.toString()));
        assertNull(board.get(D4));

        whitePawn.moveTo(D4);

        assertNull(board.get(INITIAL_POSITION));
        assertNotNull(board.get("D4"));
    }

    @Test
    public void getAvailablePositionsFromD3Test() {
        final Set<BoardPosition> pos = whitePawn
                .moveTo("D3")
                .getAvailablePositions();
        assertEquals(1, pos.size());
    }

    @Test
    public void getAvailablePositionsFromD8Test() {
        final Set<BoardPosition> pos = whitePawn
                .moveTo(D3)
                .moveTo(D4)
                .moveTo(D5)
                .moveTo(D6)
                .moveTo(D7)
                .moveTo(D8)
                .getAvailablePositions();
        assertEquals(0, pos.size());
    }
}

File: core-kotlin/src/test/kotlin/chessgame/domain/model/PawnWhiteTest.kt
package chessgame.domain.model

import chessgame.domain.Piece
import chessgame.domain.PieceFactory.Companion.createWhitePawn
import chessgame.domain.model.BoardPosition.*
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test


class PawnWhiteTest {

    companion object {
        private val INITIAL_POSITION = D2
    }

    private var board: Board? = null
    private var whitePawn: Piece? = null

    @Before
    fun setUp() {
        board = Board()
        board!!.init()
        whitePawn = createWhitePawn(board!!, INITIAL_POSITION)
    }

    @Test
    fun createInstanceTest() = assertTrue(whitePawn is Pawn)

    @Test
    fun whiteColorTest() = assertEquals(PieceColor.WHITE, whitePawn!!.color)

    @Test
    fun initPositionTest() {
        assertEquals(INITIAL_POSITION, whitePawn!!.position)
        assertEquals(whitePawn, board!![INITIAL_POSITION])
    }

    @Test
    fun moveToD4Test() {
        assertNotNull(board!![INITIAL_POSITION.toString()])
        assertNull(board!![D4])

        whitePawn!!.moveTo(D4)

        assertNull(board!![INITIAL_POSITION])
        assertNotNull(board!!["D4"])
    }

    @Test
    fun getAvailablePositionsFromD3Test() = assertEquals(1, whitePawn!!
            .moveTo("D3")
            .getAvailablePositions()
            .size)

    @Test
    fun getAvailablePositionsFromD8Test() = assertEquals(0, whitePawn!!
            .moveTo(D3)
            .moveTo(D4)
            .moveTo(D5)
            .moveTo(D6)
            .moveTo(D7)
            .moveTo(D8)
            .getAvailablePositions()
            .size)
}